web/src/reducers/budget.js

```diff
@@ -0,0 +1,149 @@
+import { UPDATE_ACTIVITY } from '../actions/activities';
+import { GET_APD_SUCCESS } from '../actions/apd';
+import { YEAR_OPTIONS } from '../util';
+
+const getFundingSourcesByYear = (years = YEAR_OPTIONS) => ({
+  ...years.reduce(
+    (o, year) => ({
+      ...o,
+      [year]: {
+        total: 0,
+        federal: 0,
+        state: 0
+      }
+    }),
+    {}
+  ),
+  total: {
+    total: 0,
+    federal: 0,
+    state: 0
+  }
+});
+
+const expenseTypes = () => ({
+  statePersonnel: getFundingSourcesByYear(),
+  contractors: getFundingSourcesByYear(),
+  expenses: getFundingSourcesByYear(),
+  combined: getFundingSourcesByYear()
+});
+
+const initialState = () => ({
+  combined: getFundingSourcesByYear(),
+  hie: expenseTypes(),
+  hit: expenseTypes(),
+  mmis: expenseTypes(),
+  years: YEAR_OPTIONS
+});
```

Comment:
These are all functions so that when I'm building the budget table from scratch, all of these properties point to unique objects.  If I'd done something like this, they'd point to the same instances:

```js
const expenseTypes = {
  statePersonnel: {},
  contractors: {},
  expenses: {},
  combined: {}
};

const initialState = {
  combined: {},
  hie: expenseTypes,
  hit: expenseTypes,
  mmis: expenseTypes
}

initialState.hie.statePersonnel === initialState.mmis.statePersonnel // true
```

That's totally fine for a pure "initial state" object, but it doesn't work when we're actually trying to fill it in with data.  And rather than try to figure out which piece of budget data changed, it made way more sense to me to just build the whole table each time.  So...  Here we are.  🙂

Reply:
👍